node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeRepository.java

```diff
@@ -468,28 +468,60 @@ private Node move(Node node, Node.State toState, Agent agent, Optional<String> r
     }
 
     /**
-     * Removes a node. A node must be in a legal state before it can be removed.
+     * Removes all the nodes that are children of hostname before finally removing the hostname itself.
+     *
+     * @return List of all the nodes that have been removed
      */
-    public void remove(String hostname) {
-        Node nodeToRemove = getNode(hostname).orElseThrow(() ->  new NotFoundException("No node with hostname \"" + hostname + '"'));
-        List<Node.State> legalStates = dynamicAllocationEnabled() ?
-                Arrays.asList(Node.State.provisioned, Node.State.failed, Node.State.parked, Node.State.dirty) :
-                Arrays.asList(Node.State.provisioned, Node.State.failed, Node.State.parked);
-
-        if (! legalStates.contains(nodeToRemove.state())) {
-            throw new IllegalArgumentException("Can only remove node from following states: " +
-                    legalStates.stream().map(Node.State::name).collect(Collectors.joining(", ")));
+    public List<Node> removeRecursively(String hostname) {
+        Node node = getNode(hostname).orElseThrow(() -> new NotFoundException("No node with hostname \"" + hostname + '"'));
+
+        try (Mutex lock = lockUnallocated()) {
+            List<Node> removed = node.type() != NodeType.host ?
+                    new ArrayList<>() :
+                    getChildNodes(hostname).stream()
+                            .filter(child -> allowedToRemove(child, true))
+                            .collect(Collectors.toList());
+
+            if (allowedToRemove(node, false)) removed.add(node);
+            db.removeNodes(removed);
+
+            return removed;
+        } catch (RuntimeException e) {
+            throw new IllegalArgumentException("Failed to delete " + node.hostname(), e);
         }
+    }
+
+    /**
+     * Allowed to a node delete if:
+     *  Non-docker-container node: iff in state provisioned|failed|parked
+     *  Docker-container-node:
+     *    If only removing the container node: node in state ready
+     *    If also removing the parent node: child is in state provisioned|failed|parked|ready
+     */
+    private boolean allowedToRemove(Node nodeToRemove, boolean deletingAsChild) {
```

Comment:
Could we change the API and name of this function? Looks like the method should actually return void but throw exception on error. checkRemovalIsAllowed? verifyRemovalIsAllowed? ...

Reply:
Can change the name, but the API is this way so that it can easily be used in stream `filter()` while also throwing straight to REST API handler in case we try to perform invalid operation.